File: MealTracker/TrackerTableViewController.m
//
//  TrackerTableViewController.m
//  MealTracker
//
//  Created by Will Hindenburg on 7/25/12.
//  Copyright (c) 2012 __MyCompanyName__. All rights reserved.
//

#import "TrackerTableViewController.h"
#import "RetrieveMealHistoryService.h"
#import "MealEaten.h"
#import "Meal.h"
#import "DeleteMealEatenService.h"

@import BaseClasses;

@interface TrackerTableViewController ()
@property (strong, nonatomic) User *user;
@property (strong, nonatomic) NSArray *dataSource;
@end

@implementation TrackerTableViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
	
    self.navigationItem.leftBarButtonItem = self.editButtonItem;
}

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    [self retrieveMealHistory];
    
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    self.user = [User persistentUserObject];
}

- (void)retrieveMealHistory {
    [super showActivityIndicatorAnimated:YES];
    RetrieveMealHistoryService *service = [RetrieveMealHistoryService new];
    [service loadMealHistoryBasedOnUser:self.user withSuccessBlock:^(NSArray *data) {
        self.dataSource = data;
        [self.tableView reloadData];
        [super hideActivityIndicatorAnimated:YES];
    } andError:^(NSError *error) {
        [super hideActivityIndicatorAnimated:YES];
        [super showError:error withRetryBlock:^{
            [self retrieveMealHistory];
        }];
    }];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *CellIdentifier = @"Date Eaten Cell";
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier];
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:CellIdentifier];
    }

    MealEaten *mealEaten = [self mealEatenForIndexPath:indexPath];
    
    NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
    [dateFormatter setDateStyle:NSDateFormatterMediumStyle];
    [dateFormatter setTimeStyle:NSDateFormatterShortStyle];
    
    NSString *formattedDateString = [dateFormatter stringFromDate:mealEaten.dateEaten];
    NSString *fullString = [[NSString alloc] initWithFormat:@"%@ - %@ Points", formattedDateString, [mealEaten.meal.weightWatchersPlusPoints stringValue]];
    cell.detailTextLabel.text = fullString;
    cell.textLabel.text = mealEaten.meal.name;
    
    return cell;
}

- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section {
    return 45.0f;
}

#pragma mark - UITableViewDataSource

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return self.dataSource.count;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    NSDictionary *dataSourceForSection = [self.dataSource objectAtIndex:section];
    NSArray *array = [dataSourceForSection valueForKey:@"meals"];
    return array.count;
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    NSDictionary *dataSourceForSection = [self.dataSource objectAtIndex:section];
    NSDate *date = [dataSourceForSection valueForKey:@"month"];
    NSArray *array = [dataSourceForSection valueForKey:@"meals"];
    
    NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
    dateFormatter.dateStyle = NSDateFormatterMediumStyle;
    dateFormatter.timeStyle = NSDateFormatterNoStyle;
    
    NSInteger points = 0;
    for (MealEaten *dateEaten in array)
    {
        points += [dateEaten.meal.weightWatchersPlusPoints integerValue];
    }
    NSInteger pointsLeft = [self.user.pointsPerWeek integerValue] - points;
    NSString *sectionTitle = [dateFormatter stringFromDate:date];
    NSString *fullSectionTitle = [[NSString alloc] initWithFormat:@"%@ - %li Points Left", sectionTitle, (long)pointsLeft];
    return fullSectionTitle;
    
}

- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        [super showActivityIndicatorAnimated:YES];
        MealEaten *meal = [self mealEatenForIndexPath:indexPath];
        [[DeleteMealEatenService new] removeMealEaten:meal withSuccessBlock:^{
            NSMutableArray *mutableCapsuleArray = [self dataSourceArrayForSection:indexPath.section];
            [mutableCapsuleArray removeObjectAtIndex:indexPath.row];
            [self.tableView reloadSections:[NSIndexSet indexSetWithIndex:indexPath.section] withRowAnimation:UITableViewRowAnimationAutomatic];
            [super hideActivityIndicatorAnimated:YES];
        } andError:^(NSError *error) {
            [super hideActivityIndicatorAnimated:YES];
            [super showError:error withRetryBlock:^{
                [self tableView:tableView commitEditingStyle:editingStyle forRowAtIndexPath:indexPath];
            }];
        }];
    }
}

- (NSMutableArray *)dataSourceArrayForSection:(NSInteger)section {
    NSDictionary *dataSourceForSection = [self.dataSource objectAtIndex:section];
    NSMutableArray *array = [dataSourceForSection valueForKey:@"meals"];
    return array;
}

- (MealEaten *)mealEatenForIndexPath:(NSIndexPath *)indexPath {
    NSDictionary *dataSourceForSection = [self.dataSource objectAtIndex:indexPath.section];
    NSArray *array = [dataSourceForSection valueForKey:@"meals"];
    MealEaten *mealEaten = [array objectAtIndex:indexPath.row];
    return mealEaten;
}

@end

File: MealTracker/TrackerTableViewController.swift
//
//  TrackerTableViewController.swift
//  MealTracker
//
//  Created by William Hindenburg on 9/1/15.
//
//

import UIKit
import BaseClassesSwift

class TrackerTableViewController2: MealBaseTableViewController {

    var user = User.persistentUserObject()
    var dataSource:Array <Dictionary<String, AnyObject>> = [[String: AnyObject]]()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.navigationItem.leftBarButtonItem = self.editButtonItem
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        self.user = User.persistentUserObject();
        self.retrieveMealHistory()
    }
    
    func retrieveMealHistory() {
        super.showActivityIndicator(animated: true)
        RetrieveMealHistoryService().loadMealHistoryBased(on: self.user, withSuccessBlock: { (dataSource) -> Void in
            self.dataSource = dataSource as Array<Dictionary<String, AnyObject>>
            self.tableView.reloadData()
            super.hideActivityIndicator(animated: true)
        }) { (error) -> Void in
            super.hideActivityIndicator(animated: true)
            super.showError(error, withRetry: { () -> Void in
                self.retrieveMealHistory()
            })
        }
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cellIdentifier = "Date Eaten Cell"
        var cell = tableView.dequeueReusableCell(withIdentifier: cellIdentifier)
        if cell == nil {
            cell = UITableViewCell(style: UITableViewCellStyle.default, reuseIdentifier: cellIdentifier)
        }
        
        if let mealEaten = self.mealEatenForIndexPath(indexPath) {
            let dateFormatter = DateFormatter()
            dateFormatter.dateStyle = DateFormatter.Style.medium
            dateFormatter.timeStyle = DateFormatter.Style.short
            
            let formattedDateString = dateFormatter.string(from: mealEaten.dateEaten as Date)
            let fullString = formattedDateString + " - " + mealEaten.meal.weightWatchersPlusPoints.stringValue + " Points"
            cell?.detailTextLabel?.text = fullString
            cell?.textLabel?.text = mealEaten.meal.name
        }
        return cell!

    }
    
    override func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        return 45.0
    }
    
    override func numberOfSections(in tableView: UITableView) -> Int {
        return self.dataSource.count
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let dataSourceForSection = self.dataSource[section]
        if let array = dataSourceForSection["meals"] as? Array<MealEaten> {
            return array.count
        } else {
            return 1
        }
    }
    
    override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        let dataSourceForSection = self.dataSource[section]
        if let date = dataSourceForSection["month"] as? Date {
            if let array = dataSourceForSection["meals"] as? Array<MealEaten> {
                let dateFormatter = DateFormatter()
                dateFormatter.dateStyle = DateFormatter.Style.medium
                dateFormatter.timeStyle = DateFormatter.Style.none
                
                var points = 0
                for dateEaten in array {
                    points += dateEaten.meal.weightWatchersPlusPoints.intValue
                }
                let pointsLeft = self.user.pointsPerWeek.intValue - points
                let sectionTitle = dateFormatter.string(from: date)
                let fullSectionTitle = sectionTitle + " - " + String(pointsLeft) + " Points Left"
                return fullSectionTitle
            }
        }
        return ""
    }
    
    override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCellEditingStyle, forRowAt indexPath: IndexPath) {
        if editingStyle == UITableViewCellEditingStyle.delete {
            super.showActivityIndicator(animated: true)
            let meal = self.mealEatenForIndexPath(indexPath)
            DeleteMealEatenServiceSwift().removeMealEaten(meal!, successBlock: { () -> Void in
                var mutableCapsuleArray = self.dataSourceArrayForSection(indexPath.section)
                mutableCapsuleArray.remove(at: indexPath.row)
                self.tableView.reloadSections(IndexSet(integer: indexPath.section), with: UITableViewRowAnimation.automatic)
                super.hideActivityIndicator(animated: true)
            }, errorBlock: { (error) -> Void in
                super.hideActivityIndicator(animated: true)
                super.showError(error, withRetry: { () -> Void in
                    self.tableView(tableView, commit:editingStyle , forRowAt: indexPath)
                })
            })
        }
    }
    
    func dataSourceArrayForSection(_ section:Int) -> Array<Dictionary<String, AnyObject>> {
        let dataSourceForSection = self.dataSource[section]
        if let array = dataSourceForSection["meals"] as? Array<Dictionary<String,AnyObject>> {
            return array
        } else {
            return [[String: AnyObject]]()
        }
    }
    
    func mealEatenForIndexPath(_ indexPath:IndexPath) -> MealEaten? {
        let dataSourceForSection = self.dataSource[indexPath.section]
        if let array = dataSourceForSection["meals"] as? Array<AnyObject> {
            if let mealEaten = array[indexPath.row] as? MealEaten {
                return mealEaten
            } else {
                return nil
            }
        } else {
            return nil
        }
        
    }

}
